### PR TITLE
Add ability to toggle between single and multi-threaded execution

### PR DESCRIPTION
**Pull request by peppy**
Toggle with Ctrl+Alt+F7

Note that the execution rate becomes dependent on the window loop at this point, which is generally locked to whatever the WM or OS is providing us with. Our local threads still attempt to keep their relative execution rates, meaning there is potential for sub-optimal frame times depending on frame limiter setting (using unlimited or vsync is recommended for consistency).

**This is intended only for profiling/benchmarking purposes.** It is a barebones implementation with missing safeties (for example, the thread checks are bypassed, rather than checking for input thread). I mainly just want this on appveyor for easier testability.

That said, I won't decline reviews or discussion on the direction of implementation.

Road forward:
- Test performance / heat / energy metrics on mobile devices
- Short term: if a huge improvement is seen on mobile, consider hardcoding enabled (and remove toggle logic if it seen as too complex)
- Compare performance on single core desktop (for fair comparisons I intend on creating a BenchmarkDotNet game loop for this)

Update:

- Initial benchmarks show a 60-70% reduction in CPU usage (iPhone SE / SampleGame).
- Blocking audio operations will cause a hard freeze due to waiting on a threaded operation that will never be reached.

.

- [x] Depends on https://github.com/ppy/osu-framework/pull/3313
- [x] Depends on https://github.com/ppy/osu-framework/pull/3314
- [x] Depends on https://github.com/ppy/osu-framework/pull/3317

### CODE REVIEW

**Comment by peppy** [comment]
This is ready for review. Now dependent on https://github.com/ppy/osu-framework/pull/3313.

File: osu.Framework/Platform/ThreadRunner.cs [hunk 1]
```diff
@@ -0,0 +1,205 @@
...
+                    break;
+                }
+
+                case ExecutionMode.SingleThread:
+                {
+                    // switch to single-threaded.
+                    foreach (var t in Threads)
+                    {
+                        t.MakeCurrent();
+
```
Comment (smoogipoo) [inline osu.Framework/Platform/ThreadRunner.cs]:
```suggestion
```

File: osu.Framework/Threading/UpdateThread.cs [hunk 2]
```diff
@@ -5,21 +5,35 @@
...
-        public UpdateThread(Action onNewFrame)
+        private readonly DrawThread drawThread;
+
+        public UpdateThread(Action onNewFrame, DrawThread drawThread)
             : base(onNewFrame, "Update")
         {
+            this.drawThread = drawThread;
+        }
+
+        protected override void OnInitialize()
```
Comment (smoogipoo) [inline osu.Framework/Threading/UpdateThread.cs]:
```suggestion
        protected sealed override void OnInitialize()
```

File: osu.Framework/Threading/UpdateThread.cs [hunk 3]
```diff
@@ -5,21 +5,35 @@
...
+            if (ThreadSafety.ExecutionMode != ExecutionMode.SingleThread)
+            {
+                //this was added due to the dependency on GLWrapper.MaxTextureSize begin initialised.
+                drawThread?.WaitUntilInitialized();
+            }
         }
 
         public override bool IsCurrent => ThreadSafety.IsUpdateThread;
 
         internal override void MakeCurrent()
```
Comment (smoogipoo) [inline osu.Framework/Threading/UpdateThread.cs]:
```suggestion
        internal sealed override void MakeCurrent()
```

File: osu.Framework/Threading/DrawThread.cs [hunk 4]
```diff
@@ -5,24 +5,54 @@
...
+        public DrawThread(Action onNewFrame, GameHost host)
             : base(onNewFrame, "Draw")
         {
+            this.host = host;
         }
 
         public override bool IsCurrent => ThreadSafety.IsDrawThread;
 
-        internal override void MakeCurrent()
+        protected override void OnInitialize()
```
Comment (smoogipoo) [inline osu.Framework/Threading/DrawThread.cs]:
```suggestion
        protected sealed override void OnInitialize()
```

File: osu.Framework/Threading/DrawThread.cs [hunk 5]
```diff
@@ -5,24 +5,54 @@
...
+        }
+
+        internal sealed override void MakeCurrent()
         {
             base.MakeCurrent();
+
             ThreadSafety.IsDrawThread = true;
         }
 
+        protected override void Cleanup()
```
Comment (smoogipoo) [inline osu.Framework/Threading/DrawThread.cs]:
```suggestion
        protected sealed override void Cleanup()
```

File: osu.Framework/Threading/AudioThread.cs [hunk 6]
```diff
@@ -23,6 +23,7 @@ public AudioThread()
         internal override void MakeCurrent()
```
Comment (smoogipoo) [inline osu.Framework/Threading/AudioThread.cs]:
```suggestion
        internal sealed override void MakeCurrent()
```

**Pull request merged**